### ISSUE TITLE
endpoint to request email confirmation to be resent

### ISSUE THREAD

**Issue by guyjacks** [Tivix/django-rest-auth#343]
Is there endpoint that can be used to request that the email verification be resent?

**Comment by dkzlv** [comment]
@guyjacks AFAIK, there's no such thing here.

If you're using allauth (which you should probably do), I believe, all you need to do is to make a simple view, which will use `send_email_confirmation` from `allauth.account.utils` module.

**Comment by mbaragiola** [comment]
`django-allauth` says that when you try to login to an unconfirmed email account, it fails to login and resends the confirmation email. I'm not getting such results in `django-rest-auth`.

Anyone knows what's up?

@guyjacks @dkzlv 

Thanks.

**Comment by chdsbd** [comment]
I would find such an endpoint useful. It might be better to have it part of rest auth instead of everyone having to roll their own.

**Comment by ruathudo** [comment]
How is the status of this?
It would be nice with /send_verification_email endpoint so we can use whenever use want to resend.
I see this is usable in most of the case.

**Comment by Shelagh-Lewins** [comment]
Thanks! Unfortunately, I fail at the first step. I get an error "ImportError: cannot import name 'action'"

```
from rest_framework.decorators import action

class UserListView(generics.ListCreateAPIView):
    queryset = models.CustomUser.objects.all()
    serializer_class = serializers.UserSerializer

    @action(
        detail=True,
        methods=['patch'],
        url_path='resend',
        #permission_classes=[permissions.EmailAddressResendPermission]
    )
    def resend(self, request, pk=None):
        print('resend')
```
I'm not using Docker and I can't see why the import fails.
djangorestframework==3.7.7

**Comment by Shelagh-Lewins** [comment]
Although 'action' [is in the docs](https://www.django-rest-framework.org/api-guide/routers/#routing-for-extra-actions) it doesn't seem to exist in the rest-framework source code. I found and tried this instead:

```
from django.views.decorators.csrf import csrf_exempt
from django.utils.decorators import method_decorator
from rest_framework.decorators import detail_route

class UserListView(generics.ListCreateAPIView):
    queryset = models.CustomUser.objects.all()
    serializer_class = serializers.UserSerializer

    @method_decorator(csrf_exempt)
    @detail_route(
        detail=True,
        methods=['patch'],
        url_path='resend',
    )
    def resend(self, request, pk=None):
        print('resend')
```

but I still get the error Forbidden (CSRF cookie not set.): /api/v1/rest-auth/user/resend even though this is calling my custom user view (which works!) not the allauth endpoint.

**Comment by mbaragiola** [comment]
I can't tell for sure seeing just that code, but sounds like you're trying to open some HTML view? CSRF shouldn't be a problem when using JSON RESTful APIs @Shelagh-Lewins

**Comment by Shelagh-Lewins** [comment]
Yes, that's what is so weird! The UserListView is a simple endpoint to return a list of users in DRF, it works fine, but the detail_view gives CSRF error on post. I also tried creating a new APIView but that also gives CSRF error on post. I wish I could find any full working code example or tutorial.

For now, I have given up. Absolutely the only way I can make this work is to define a custom get on an endpoint. Which is wrong, but I can't spend any more time on what should be a standard, simple feature. I cannot figure out what I've done wrong.

**Comment by Shelagh-Lewins** [comment]
This section of the [documentation](https://www.django-rest-framework.org/topics/ajax-csrf-cors/) seems to say that you *do* need to do CSRF with DRF. But I can't make it work from the [instructions](https://docs.djangoproject.com/en/2.1/ref/csrf/#ajax). So confused!

**Comment by Shelagh-Lewins** [comment]
I seem to have this working now with a custom view in my views.py:

class EmailConfirmation(APIView):
    permission_classes = [IsAuthenticated] 

    def post(self, request):
        send_email_confirmation(request, request.user)
        return Response({'message': 'Email confirmation sent'}, status=status.HTTP_201_CREATED)

It seems like if you have an error in your endpoint URL when you send the request, the server gives you the message "Forbidden (CSRF cookie not set.)" which obviously made me think the issue was with CSRF. So I think part of my issue was getting mixed up with exactly what are my API URLs.

**Comment by brunobarretofreitas** [comment]
Check my pull request #530
